Mathlib/Algebra/Group/UniqueProds.lean

```diff
@@ -196,36 +216,153 @@ attribute [to_additive UniqueSums] UniqueProds
+      exact Finset.mem_mul.mpr ⟨a0, b, ha0, hb, rfl⟩
+
+open MulOpposite in
+-- see Note [lower instance priority]
+/-- This instance asserts that if `A` has a left-cancellative multiplication, a linear order,
+  and multiplication is strictly monotone w.r.t. the first argument, then `A` has `UniqueProds`. -/
+@[to_additive _root_.Covariant.to_uniqueSums_left
+  "This instance asserts that if `A` has a left-cancellative addition, a linear order,
+  and addition is strictly monotone w.r.t. the first argument, then `A` has `UniqueSums`." ]
+instance (priority := 100) _root_.Covariant.to_uniqueProds_left {A} [Mul A] [IsLeftCancelMul A]
+    [LinearOrder A] [CovariantClass A A (Function.swap (· * ·)) (· < ·)] :
+    UniqueProds A :=
+let _ := LinearOrder.lift' (unop : Aᵐᵒᵖ → A) unop_injective
+let _ : CovariantClass Aᵐᵒᵖ Aᵐᵒᵖ (· * ·) (· < ·) :=
+{ elim := fun _ _ _ bc =>
+          have : StrictMono (unop (α := A)) := fun _ _ => id
+          mul_lt_mul_right' (α := A) bc (unop _) }
+of_mulOpposite _ Covariant.to_uniqueProds_right
+
+variable {G H : Type*} [Mul G] [Mul H]
+
+@[to_additive (attr := nontriviality, simp)]
+theorem of_subsingleton [Subsingleton G] : UniqueProds G :=
+  ⟨fun ⟨a, ha⟩ ⟨b, hb⟩ => ⟨a, ha, b, hb, UniqueMul.of_subsingleton⟩⟩
+
+open Finset
+@[to_additive]
+theorem mulHom_image_of_injective (f : H →ₙ* G) (hf : Function.Injective f) (uG : UniqueProds G) :
+    UniqueProds H := by
+  refine ⟨fun {A B} A0 B0 => ?_⟩
+  classical
+  obtain ⟨a0, ha0, b0, hb0, h⟩ := uG.uniqueMul_of_nonempty (A0.image f) (B0.image f)
+  rcases mem_image.mp ha0 with ⟨a', ha', rfl⟩
+  rcases mem_image.mp hb0 with ⟨b', hb', rfl⟩
+  exact ⟨a', ha', b', hb', (UniqueMul.mulHom_image_iff f hf).mp h⟩
+
+/-- `UniqueProd` is preserved under multiplicative equivalences. -/
+@[to_additive "`UniqueSums` is preserved under additive equivalences."]
+theorem mulHom_image_iff (f : G ≃* H) :
+    UniqueProds G ↔ UniqueProds H :=
+⟨mulHom_image_of_injective f.symm f.symm.injective, mulHom_image_of_injective f f.injective⟩
+
+@[to_additive] instance [UniqueProds G] [UniqueProds H] : UniqueProds (G × H) where
+  uniqueMul_of_nonempty {A B} hA hB := by
+    classical
+    obtain ⟨aG, hA, bG, hB, hG⟩ := uniqueMul_of_nonempty (hA.image Prod.fst) (hB.image Prod.fst)
+    rw [mem_image, ← filter_nonempty_iff] at hA hB
+    obtain ⟨aH, hA, bH, hB, hH⟩ := uniqueMul_of_nonempty (hA.image Prod.snd) (hB.image Prod.snd)
+    simp_rw [mem_image, mem_filter] at hA hB
+    refine ⟨(aG, aH), ?_, (bG, bH), ?_, fun a b ha hb he => ?_⟩
+    · obtain ⟨a, ⟨ha, rfl⟩, rfl⟩ := hA; exact ha
+    · obtain ⟨b, ⟨hb, rfl⟩, rfl⟩ := hB; exact hb
+    rw [Prod.ext_iff] at he
+    specialize hG (mem_image.mpr ⟨a, ha, rfl⟩) (mem_image.mpr ⟨b, hb, rfl⟩) he.1
+    specialize hH _ _ he.2
+    all_goals try simp_rw [mem_image, mem_filter]
+    exacts [⟨a, ⟨ha, hG.1⟩, rfl⟩, ⟨b, ⟨hb, hG.2⟩, rfl⟩, ⟨Prod.ext hG.1 hH.1, Prod.ext hG.2 hH.2⟩]
```

Comment:
may or may not be easier to prove via isomorphism with a `Pi` type indexed by `bool`.

Reply:
Hmm, I did not try much, but... is the `Pi` instance universe polymorphic?  Are the `G i` all in the same universe?  Maybe it is easier to keep the proof for a product separate.

Reply:
Oh indeed universe is an issue, so let's leave it as is.

Reply:
Here is my attempt: I am missing the "multiplicativity, and I do not know how to put that instance effectively...

```lean
@[to_additive] instance [UniqueProds G] [UniqueProds H] : UniqueProds (G × H) := by
  have : G × H ≃ ∀ x : Bool, if x then ULift.{max u_1 u_2} G else ULift.{max u_1 u_2} H := by
    refine ⟨?_, ?_, ?_, ?_⟩
    · intro (g, h) x
      cases x
      simpa only [ite_false] using { down := h }
      simpa only [ite_true] using { down := g }
    · intros x
      let xG := x true
      let xH := x false
      simp at xG xH
      exact ⟨xG.down, xH.down⟩
    · intro (g, h)
      simp
    . intro h
      ext x
      cases x
      · simp; rfl
      · simp; rfl
  apply UniqueProds.mulHom_image_iff.mp -- missing the ≃*!
```

Reply:
We usually define a function out of `Bool` using `Bool.rec`; hopefully it simplifies the proof!

Reply:
I'll give it a go, but tomorrow.  In any case, I'm not sure that aiming for shortest proof is necessarily an advantage.

I'll also try with the injection only.

Anyway, you are right that the equivalence between `Prod` and `Bool → ...` should be in mathlib.  I'll try to look for it (and prove it as well, to get some practice with universe levels).  If I cannot find it, I'll PR it.

Reply:
There is [Equiv.boolArrowEquivProd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Basic.html#Equiv.boolArrowEquivProd) but not the dependent version in the same file, and that should be pretty good evidence that it doesn't exist yet ...

Reply:
Ok, I will probably PR a version of the equivalence of
* `Prod` with factors potentially in different universes and 
* dependent function from `Bool`.

(This should be what is already in this thread, I think.)

However, likely, I will not try to force through also a typeclass assumption on the types and will keep the two proofs for `Pi` Ans `Prod` of `UniqueProds` that we currently have.

Reply:
Oh I somehow missed this reply. 

I just found [prodEquivPiFinTwo](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#prodEquivPiFinTwo) and [piFinTwoEquiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Fin.html#piFinTwoEquiv) which use `Fin 2` rather than `Bool`. A `Bool` version would feel a bit redundant. I also think the fully universe polymorphic version is not "atomic" in the sense it can be obtained from the version with both types in the same universe by composing with [Equiv.prodCongr](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Basic.html#Equiv.prodCongr) applied to [Equiv.ulift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Defs.html#Equiv.ulift), so I'd be satisfied with the less polymorphic version.

Reply:
I'm going to leave this as is: I think that the current proof is nicer than trying to force a polymorphic multiplicative equivalence with typeclasses between `Prod` and `Pi`.

Reply:
I can prove it in 11 lines https://github.com/leanprover-community/mathlib4/compare/adomani/UniqueProds_for_LW...jyxu/UniqueProds_prod_from_pi?expand=1#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1R352-R362 but to_additive strangely has trouble working.

Reply:
I'll take a look!

Reply:
The issue seems to be caused by the `inferInstance`s.

I solved the first 4 by replacing them with explicit `ULift.mul`.  I am working on figuring out the last two `inferInstance`s in `mulHom_image_of_injective`.

Reply:
There is also a combined issue that the `cases b` embeds something in the proof term that `to_additive` misinterprets.

Reply:
I have tried a little to find a way to prove the instance without messing with `to_additive`, but I was not able to.  This version causes problems *only* due to the `cases b` at the end:
```lean
@[to_additive]
instance [uG : UniqueProds G] [uH : UniqueProds H] : UniqueProds (G × H) := by
  let I : Bool → Type max u v := Bool.rec (ULift.{v} G) (ULift.{u} H)
  let ins : Mul (∀ b, I b) := @Pi.instMul Bool I (Bool.rec myMul myMul)
  let f : MulHom (G × H) (∀ b, I b) := ⟨fun x => Bool.rec ⟨x.1⟩ ⟨x.2⟩, fun x y => ?_⟩
  swap; ext b; cases b <;> rfl
  refine mulHom_image_of_injective f (fun x y he => Prod.ext ?_ ?_) ?_
  · exact ULift.up_injective (congr_fun he false)
  · exact ULift.up_injective (congr_fun he true)
  refine @instUniqueProdsForAllInstMul _ I (Bool.rec myMul myMul) ?_
  intro b
  cases b <;> -- put a `sorry` before here and `to_additive` is happy
  exact mulHom_image_of_injective ⟨ULift.down, fun _ _ => rfl⟩ ULift.down_injective inferInstance
```
It is possible that the last `inferInstace` could cause further problems, but those might be soluble by explicitly passing the instance name.

Note that the "manually" `to_additivized` version works:
```lean
example [Add G] [Add H] [UniqueSums G] [UniqueSums H] : UniqueSums (G × H) := by
  let I : Bool → Type max u v := Bool.rec (ULift.{v} G) (ULift.{u} H)
  let ins : Add (∀ b, I b) := @Pi.instAdd Bool I (Bool.rec ULift.add ULift.add)
  let f : AddHom (G × H) (∀ b, I b) := ⟨fun x => Bool.rec ⟨x.1⟩ ⟨x.2⟩, fun x y => ?_⟩
  swap; ext b; cases b <;> rfl
  refine UniqueSums.addHom_image_of_injective f (fun x y he => Prod.ext ?_ ?_) ?_
  · exact ULift.up_injective (congr_fun he false)
  · exact ULift.up_injective (congr_fun he true)
  refine @UniqueSums.instUniqueProdsForAllInstAdd _ I (Bool.rec _ _) ?_
  intro b; cases b <;>
  exact UniqueSums.addHom_image_of_injective ⟨ULift.down, fun _ _ => rfl⟩ ULift.down_injective inferInstance
```

Reply:
Thanks for the investigation! [This](https://github.com/leanprover-community/mathlib4/compare/adomani/UniqueProds_for_LW...jyxu/UniqueProds_prod_from_pi?expand=1#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1R352-R364) is a version that makes `to_additive` work. The key seems to be singling out all `MulHom`s rather than inlining them (I tried using `MulHom.mk` rather than `⟨_, _⟩` and it doesn't work). Could `to_additive` be patched to support inlined `MulHom`s?

Reply:
Great!  Do you want to push it to the PR?  Should I do it?

As for the `to_additive` question, I do not know what the answer is.  I have tried to understand how `to_additive` works, I have some vague idea, but there is always something about it that seems magic...

Reply:
Feel free to push if you like the new proof! (Note the new import and explicit universes on G and H.)

Reply:
Upon further reflection, I think that I prefer the current proof: disconnected instances for `Prod` and `Pi`.  The reason that I like the proof connecting the two is because I find it interesting how you manage to get the universe levels to line up, while also juggling the limitations of `to_additive`.  From the point of view of the PR, though, this is not really a goal!

So, unless you think differently, I would leave the proofs as they are now.